## Домашнее задание №1 ##
### Линейное пространство. Основные понятия. Часть 1 ###

**1. Исследовать на линейную зависимость:** $f_{1}(x)=e^{x}, f_{2}(x)=1, f_{3}(x)=x+1, f_{4}(x)=x-e^{x}.$

Сведем все функции к: $f_{4}(x)=f_{3}(x)-f_{2}(x)-f_{1}(x),$

следовательно $f_{4}(x)$ — линейная комбинация векторов $f_{1}(x)$, $f_{2}(x)$ и $f_{3}(x)$

следовательно все четыре функции $f_{1}(x)=e^{x}, f_{2}(x)=1, f_{3}(x)=x+1, f_{4}(x)=x-e^{x}$ линейно зависимы

**Ответ: функции линейо зависимы**

**2. Исследовать на линейную зависимость:** $f_{1}(x)=2, f_{2}(x)=x, f_{3}(x)=x^{2}, f_{4}(x)=(x+1)^{2}$

Сведем все к функции: $f_{4}(x)=f_{3}(x) + 2 \cdot f_{2}(x) + \frac{1}{2} \cdot f_{1}(x),$

следовательно все четыре функции $f_{4}(x)$,$f_{1}(x)$, $f_{2}(x)$ и $f_{3}(x)$  линейно зависимы.

**Ответ: функции линейо зависимы**

**3. Найти координаты вектора** $x = (2, 3, 5)\in \mathbb{R}^{3}$ **в базисе** $b_{1}=(0, 0, 10)$, $b_{2}=(2, 0, 0)$, $b_{3}=(0, 1, 0)$.

Приведем к стандартному базису линейного пространства $\mathbb{R}^{3}$ образуют векторы 
$e_{1}=(1, 0, 0)$, $e_{2}=(0, 1, 0)$, $e_{3}=(0, 0, 1)$.

$x=(2, 3, 5)=(2,0,0)+(0, 3, 0)+(0, 0, 5)=\frac{1}{2}\cdot(1, 0, 0)+1\cdot(2, 0, 0)+3\cdot(0, 1, 0)= \frac{1}{2}b_{1}+b_{2}+3b_{3}$

соттветвенно координатами вектора $x$ в базисе $b_{1}=(0, 0, 10)$, $b_{2}=(2, 0, 0)$, $b_{3}=(0, 1, 0)$ являются $\frac{1}{2}$, $1$, $3$.

**Ответ: координатами вектора $x$ ($\frac{1}{2}$, $1$, $3$).**

**4. Найти координаты вектора** $3x^{2}-2x+2\in\mathbb{R}^{3}[x]$:

**а) в базисе** $1$, $x$, $x^{2}$;

**б) в базисе** $x^{2}$, $x-1$, $1$.

высчитаем координаты в базисе $3x^{2}-2x+2=3e_{1}-2e_{2}+2e_{3}$

a) координаты вектора $3x^{2}-2x+2\in\mathbb{R}^{3}[x]$ в базисе $1$, $x$, $x^{2}$ = (2,-2,3)

высчитаем координаты в базисе $3x^{2}-2x+2=3e_{1}-2e_{2}+2e_{3} = 3\cdot x^{2}-2\cdot(x-1)+2\cdot1 =3x^{2}-2x$

б) координаты вектора $3x^{2}-2x+2\in\mathbb{R}^{3}[x]$ в базисе $x^{2}$, $x-1$, $1$ = (3,-2,0)

**Ответ: а) (2,-2,3) б) (3,-2,0)**

**5. Установить, является ли линейным подпространством:**

**а) совокупность всех векторов трехмерного пространства, у которых по крайней мере одна из первых двух координат равна нулю;**

**б) все векторы, являющиеся линейными комбинациями данных векторов**  $\{u_{1}, u_{2}, \ldots, u_{n}\}$.

Возьмем два вектора трехмерного пространства, у которых по крайней мере одна из первых двух координат равна нулю:

$a_{1}=\{0, a, b\},$
$a_{2}=\{с, 0, d\}$

Проверим и мы увидим, что 
 
$(0, a, b)+(c, 0, d)=(c, a, b+d),$

У этого вектора одна из первых двух координат не всегда будет равна нулю, а значит данное множество не является подпространством линейного пространства $\mathbb{R}^{3}$.

б) Возьмем два вектора, являющиеся линейными комбинациями векторов $\{u_{1}, u_{2}, \ldots, u_{n}\}$ домножив на коэффициенты $\lambda$ и $\mu$:

$a_{1}=\lambda_{1}u_{1}+\lambda_{2}u_{2}+...+\lambda_{n}u_{n},$ <br>
$a_{2}=\mu_{1}u_{1}+\mu_{2}u_{2}+...+\mu_{n}u_{n},$

Проведя проверку, мы увидим, что 
 
$\lambda_{1}u_{1}+\lambda_{2}u_{2}+...+\lambda_{n}u_{n}+\mu_{1}u_{1}+\mu_{2}u_{2}+...+\mu_{n}u_{n}=(\lambda_{1}+\mu_{1})u_{1}+(\lambda_{2}+\mu_{2})u_{2}+...(\lambda_{n}+\mu_{n})u_{n},$

$\alpha\cdot (\lambda_{1}u_{1}+\lambda_{2}u_{2}+...+\lambda_{n}u_{n})=\alpha\lambda_{1}u_{1}+\alpha\lambda_{2}u_{2}+...+\alpha\lambda_{n}u_{n},$

Полученные вектора принадлежат указанному множеству векторов, то есть это множество является подпространством линейного пространства.

**Ответ: а) нет б) да.**

### Линейное пространство. Основные понятия. Часть 2 ###

**1. Найти скалярное произведение векторов $x, y \in \mathbb{R}$:**

**а) $x=(0,-3, 6),~y=(-4, 7, 9);$**

**б) $x=(7, -4, 0, 1),~y=(-3, 1, 11, 2).$**

а) $(x,y)=0\cdot-4-3\cdot7+6\cdot9=33$

In [1]:
import numpy as np
x = np.array([0, -3, 6])
y = np.array([-4, 7, 9])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')

Скалярное произведение x и y равно 33


б) $(x,y)=7\cdot-3-4\cdot1+0\cdot11+1\cdot2=-23$

In [2]:
x = np.array([7, -4, 0, 1])
y = np.array([-3, 1, 11, 2])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')

Скалярное произведение x и y равно -23


**Ответ: а) 33; б) -23.**

**2. Найти нормы векторов  (4,2,4)  и  (12,3,4)  и угол между ними.**

Найдем евклидову норму вектора $x=(4, 2, 4)$:

$\left\|x\right\|_{2} = \sqrt{4^{2}+2^{2}+4^{2}}=\sqrt{36}=6.$

In [3]:
x = np.array([4, 2, 4])
print(f'Евклидова норма вектора x равна {np.linalg.norm(x)}')

Евклидова норма вектора x равна 6.0


Найдем евклидову норму для второго вектора $y=(12, 3, 4)$:

$\left\|y\right\|_{2} = \sqrt{12^{2}+3^{2}+4^{2}}=13.$

In [4]:
y = np.array([12, 3, 4])
print(f'Евклидова норма вектора y равна {np.linalg.norm(y)}')

Евклидова норма вектора y равна 13.0


Найдем скалярное произведение векторов $(x,y)=4\cdot12+2\cdot3+4\cdot4=70$

In [5]:
print(f'Скалярное произведение векторов x и y равно {np.dot(x, y)}')

Скалярное произведение векторов x и y равно 70


Косинус угла $\varphi \in [0, \pi]$ между векторами определяется: $cos\varphi = \frac{(x,y)}{\left \|x\right \|\cdot\left \|y\right \|}$

Расчитаем $cos\varphi = \frac{70}{6\cdot13} = \frac{70}{78} \Rightarrow \varphi = \arccos\frac{70}{78} \approx 0,45687$

In [6]:
from math import acos
print(f'Угол между векторами x и y равен {acos(70/78)} рад.')

Угол между векторами x и y равен 0.45687400353266105 рад.


**Ответ: 6; 13; ~0.45687 рад.**

**3. Будет ли линейное пространство евклидовым, если за скалярное произведение принять:**

**а) произведение длин векторов;**

**б) утроенное обычное скалярное произведение векторов?**

Евклидово пространство в изначальном смысле — это пространство, свойства которого описываются аксиомами евклидовой геометрии.

1) $(x,y)=(y,x);$<br>
2) $(\lambda x, y)=\lambda(x, y);$<br>
3) $(x_{1}+x_{2},y)=(x_{1},y)+(x_{2},y);$<br>
4) $(x, x)\geq0$, причем $(x,x)=0\Leftrightarrow x=0.$

a) Скалярное произведение векторов определено как: $(x,y)=\left\|x\right\| \cdot \left\|y\right\|$

$(y,x)=\left\|y\right\| \cdot \left\|x\right\|=(x,y)$, т.е. первая аксиома выполняется

$(\lambda x,y)=\left\|\lambda x\right\| \cdot \left\|y\right\|= |\lambda|\left\|x\right\|\cdot \left\|y\right\|=|\lambda|\cdot(x,y)$, т.е. при $\lambda < 0$ вторая аксиома не выполняется, значит линейное пространство **не является** евклидовым.

б) Утроенное скалярное произведение выглядит так: $(x,y)=3 \cdot (x,y) = 3 \cdot (a_{1}b_{1}+a_{2}b_{2}+...+a_{n}b_{n})$

Очевидно что аксиомы выполняются для этого условия, значит линейное пространство является евклидовым.

**Ответ: а) не является; б) является.**

**4. Какие из нижеперечисленных векторов образуют ортонормированный базис в линейном пространстве $\mathbb{R}^{3}$:**

**а) $(1,0,0),(0,0,1);$**

**б) $(1/\sqrt{2},-1/\sqrt{2},0),(1/\sqrt{2},1/\sqrt{2},0), (0,0,1);$**

**в) $(1/2, -1/2, 0), (0, 1/2, 1/2), (0,0,1);$**

**г) $(1,0,0),(0,1,0),(0,0,1)?$**

В конечномерном евклидовом пространстве базис $\{e_{1},e_{2},...,e_{n}\}$ называется ортонормированным, если 

### $(e_{i}, e_{j})=0$ $\forall$ $i\neq j$ и $(e_{i},e_{i})=1$ $\forall$ $i\in[1, n].$ ###

In [7]:
x = np.array([1, 0, 0])
y = np.array([0, 0, 1])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')
print(f'Скалярное произведение x и x равно {np.dot(x, x)}')
print(f'Скалярное произведение y и y равно {np.dot(y, y)}')

Скалярное произведение x и y равно 0
Скалярное произведение x и x равно 1
Скалярное произведение y и y равно 1


а) Условия выполняются, вектора образуют ортонормированный базис в линейном пространстве $\mathbb{R}^{3}$.

In [8]:
x = np.array([1/np.sqrt(2),-1/np.sqrt(2),0])
y = np.array([1/np.sqrt(2),1/np.sqrt(2),0])
z = np.array([0,0,1])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')
print(f'Скалярное произведение x и z равно {np.dot(x, z)}')
print(f'Скалярное произведение y и z равно {np.dot(y, z)}')

print(f'Скалярное произведение x и x равно {np.dot(x, x)}')
print(f'Скалярное произведение y и y равно {np.dot(y, y)}')
print(f'Скалярное произведение z и z равно {np.dot(z, z)}')

Скалярное произведение x и y равно 0.0
Скалярное произведение x и z равно 0.0
Скалярное произведение y и z равно 0.0
Скалярное произведение x и x равно 0.9999999999999998
Скалярное произведение y и y равно 0.9999999999999998
Скалярное произведение z и z равно 1


б) Условия выполняются, вектора образуют ортонормированный базис в линейном пространстве $\mathbb{R}^{3}$.

In [9]:
x = np.array([1/2, -1/2, 0])
y = np.array([0, 1/2, 1/2])
z = np.array([0, 0, 1])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')
print(f'Скалярное произведение x и z равно {np.dot(x, z)}')
print(f'Скалярное произведение y и z равно {np.dot(y, z)}')
print(f'Скалярное произведение x и x равно {np.dot(x, x)}')
print(f'Скалярное произведение y и y равно {np.dot(y, y)}')
print(f'Скалярное произведение z и z равно {np.dot(z, z)}')

Скалярное произведение x и y равно -0.25
Скалярное произведение x и z равно 0.0
Скалярное произведение y и z равно 0.5
Скалярное произведение x и x равно 0.5
Скалярное произведение y и y равно 0.5
Скалярное произведение z и z равно 1


в) Условия не выполняются, вектора не образуют ортонормированный базис в линейном пространстве $\mathbb{R}^{3}$.

In [10]:
x = np.array([1, 0, 0])
y = np.array([0, 1, 0])
z = np.array([0, 0, 1])
print(f'Скалярное произведение x и y равно {np.dot(x, y)}')
print(f'Скалярное произведение x и z равно {np.dot(x, z)}')
print(f'Скалярное произведение y и z равно {np.dot(y, z)}')
print(f'Скалярное произведение x и x равно {np.dot(x, x)}')
print(f'Скалярное произведение y и y равно {np.dot(y, y)}')
print(f'Скалярное произведение z и z равно {np.dot(z, z)}')

Скалярное произведение x и y равно 0
Скалярное произведение x и z равно 0
Скалярное произведение y и z равно 0
Скалярное произведение x и x равно 1
Скалярное произведение y и y равно 1
Скалярное произведение z и z равно 1


Условия выполняются, вектора образуют ортонормированный базис в линейном пространстве $\mathbb{R}^{3}$.

**Ответ: а) образуют; б) образуют; в) не образуют; г) образуют.**